### ISSUE TITLE
Hacker's Keyboard Easily Destroys Documents!

### ISSUE THREAD

**Issue by DavidPesta** [klausw/hackerskeyboard#487]
**EDIT: Ctrl-z doesn't work! Please read things closely before commenting!**

You must move the position of the Ctrl key to be more distant from the A key. At least once a week, my finger hits between the Ctrl and the A and selects all of the text in my document for my next rapid keystroke to clear and destroy the entire document. **Sadly, following up with Ctrl-z does not bring it back.** (Edit: Emphasis added.) This makes the Hacker's Keyboard useless for all applications that do not keep an archive of my work because my documents keep being destroyed like weapon factories in Europe during WW2.

One possibility is to swap the position of the Ctrl key and the Esc key, or at least provide us with that option. The keyboard is just not reliable for most purposes when you risk losing your documents.

I considered a Ctrl-A select-all failsafe, but that won't work if the failsafe message is positioned over the keyboard in any way because your next keystroke might be positioned right under the OK button that pops up, and with a fully selected document might come another keystroke that destroys it all. You would have to make sure the failsafe message appears above the keyboard where that couldn't happen.

If you don't swap the keys and the failsafe idea is no good, you should at least give us the option to turn off the Ctrl-A select-all functionality altogether.

There might be other ideas. But this absolutely must be solved. It is a critical shortcoming with an otherwise marvelous keyboard.

**EDIT: Ctrl-z doesn't work! Please read things closely before commenting!**

**Comment by TPS** [comment]
If you can believe it, this has already been filed many years ago, & is now imported into #43.…  I really appreciate the re-filed issue, though, as this is 1 of my pet peeves, also!

**Comment by DavidPesta** [comment]
I read these discussions (particularly #43 and #172) and I am astonished about the long running decision to allow so many valuable documents of so many users to be destroyed in favor of not swapping Ctrl and Esc "to avoid confusing people who got used to the current way of doing things" in the face of an argument that the current position of those keys compared to ordinary keyboards was already confusing to people.

Instead of imposing the mild discomfort of retraining people about the position of Ctrl and Esc (which is a more natural layout for newcomers anyway), entire documents for these people are being completely destroyed on a regular basis. This will go down in history as a textbook case study that illustrates very vividly when it is of paramount importance for a software maintainer to immediately apply an uncomfortable change to fix an egregious design flaw and I hope that this conversation is preserved forever for that purpose.

This is why business people rule the world and not developers.

**Comment by TPS** [comment]
The _experimental_ 5-row keyboard option _does_ move the<kbd>`Ctrl`</kbd>from next to<kbd>`a`</kbd>to kitty-corner to<kbd>`z`</kbd>, instead, so less likely to fat-finger, @ the expense of many previously-accessible symbols:

![In-situ experimental keyboard](https://lh3.googleusercontent.com/-_d6-3fC4FZc/Tv1Yr_TKSxI/AAAAAAAAAJ4/dVpWfafSzTo/w706-h1201-p-rw/IMG_20111229_222044.jpg)

**Comment by badbob001** [comment]
Where do I download the experimental version? The release version from google play still has the ctrl key next to A when in five row portrait mode (and cursor keys are also inverted -T unlike the screenshot above).

**Comment by TPS** [comment]
It's buried as part of the base download as a ___US English-only layout___ already, but needs to be enabled 2×:
 1. Get to HK's settings (@ least 3 ways to do so, but simplest is the funny-looking settings key next to<kbd>`Esc`</kbd>on typical 5-row keyboard when HK is already the default).
 2. Under the "Keyboard View Settings" header, choose "Theme and Label Settings."
 3. Toward the bottom, _after_ reading caveats in description, "Enable compact 5-row keyboard mode."
 4. Back all the way out of all HK settings, so the above is saved.
 5. Go back to main HK settings page & enable said layout under "Keyboard mode, portrait" &/or "Keyboard mode, landscape."
 6. Back all the way out again to save.
 7. Find a text field & play with _shiny, new keyboard_!

It's a reasonable layout, but I had to revert because I missed access to all the native extended symbols to which I otherwise had access!

**Comment by Referee2000** [comment]
Yes, I would love to see a settings checkbox to swap the CTRL and ESC keys of the standard keyboard.  Off by default, but allow the user to choose.  Large tablets allow two hand typing and CTRL-A is so easy to fat finger.  I would vote for a settings checkbox as the first comment suggested.  (Never had an issue with my phone, which is handheld.)

**Comment by Rhialto** [comment]
Interesting that people have these problems with the Control key, since for me (on a no-name tablet with Android 4.2.2) the Control key does not work at all for me! I think it used to work, but probably some update of something broke it.
(Also I think that the only correct position of the Control key is left to the A, like on a proper terminal such as the ADM-3a)

**Comment by DavidPesta** [comment]
@Rhialto - **Totally untrue!** Absolutely 0% of all keyboards in this Google image search have the ctrl key next to the A key. Ctrl is always on the bottom row, usually in the left corner.

https://www.google.com/search?q=computer+keyboard&tbm=isch

The Hacker's Keyboard is 1. extremely abnormal and 2. totally dangerous to my workflow! Risk of losing documents is a major software design flaw.

**Comment by Rhialto** [comment]
@DavidPesta: then you haven't seen any picture of the Lear Siegler ADM-3a keyboard. One of the first cheap ($1195) CRT terminals. It is the terminal that Vi grew up on (hence the hjkl cursor movements). Also the DEC VT-100 terminal has the CTRL left of the A, albeit with Caps Lock in between. No sensible terminal has the control key so far away in the bottom. Even the horrrible DEC LK201 keyboard (which has ESC in a stupid place) has CTRL on the same row as the A.

Of course this is less relevant for on-screen keyboards.

**Comment by DavidPesta** [comment]
@Rhialto - Your politeness is appreciated as you reminisce about old times, but this is 2016. You cannot refer to ancient technology to justify the destruction of documents by what amounts to a dangerous design flaw in an otherwise beautiful smart phone app.

**Comment by Cilencia** [comment]
I don't suppose your 2016 app supports Ctrl-Z for Undo...
Talk about a design flaw.

To weigh in on the matter, though, I typically use swipe keyboards for text input in whatever language I happen to be typing; I only use the hacker keyboard when I need special functions not available on those keyboards, Ctrl-V in particular (it's often easier to switch keyboards than get the clipboard menu to appear).

**Comment by TPS** [comment]
>  _I don't suppose your 2016 app supports Ctrl-Z for Undo…
Talk about a design flaw._

_@Cilencia Just FYI, <kbd>Ctrl</kbd>+<kbd>Z</kbd> is an **app** function, not a **keyboard** function, &, as such, works perfectly via HK when the underlying app supports it._

**Comment by Cilencia** [comment]
That was sort of my point.

**Comment by DavidPesta** [comment]
@Cilencia - I _did_ talk about the failure of Ctrl-z in my original post. Let me give you more detail on that.

Google Keep (100 million downloads)
ColorNote Notepad Notes (100 million downloads)
Keep My Notes - Notepad & Memo (1 million downloads)
Simple Notepad (1 million downloads)
Notepad (5 million downloads)

The problem is similar across these apps and the story goes something like this. A simple Ctrl-a mistake while rapidly typing will destroy the document. In many cases, the destroyed document is instantly sync'd to the app's cloud. Turning off the app's auto-sync is not an option in many cases. Universally, pressing Ctrl-z numerous times in panic or while calm does **_not_** recover the document in any of these apps.

So yes, apps still suck in 2016, but especially when used with the Hacker's Keyboard or any other virtual keyboard with "Ctrl" next to the "a".

Let's not fix thousands of broken apps. Let's just fix the Hacker's Keyboard.

**Comment by Cilencia** [comment]
While I heartily agree with you, I am not particularly optimistic for forthcoming updates to Hackers Keyboard, as Google Play lists it's last update as Dec 2013.

It appears Mr. Weidner was working on a way for users to create custom keyboards, but has moved his focus to other projects.

**Comment by TPS** [comment]
@DavidPesta Given that @Cilencia is probably correct about @KlausW, & given that [this code is under a Apache-compatible license](https://github.com/klausw/hackerskeyboard/issues/501#issuecomment-225813985), why don't you [fork](https://github.com/klausw/hackerskeyboard#fork-destination-box) & [fix it yourself](https://github.com/klausw/hackerskeyboard/network), rather than unproductively griping?  @KlausW has done a vast amount of work (when he wanted to) toward a free software, building an amazing framework.  If you do as much & contribute back, you'll have solved a problem for many. :clap:

**Comment by DavidPesta** [comment]
@TPS - Because:

1. I don't have the time or resources or perhaps ability to fix it right now

2. Somebody else who is familiar with the code might

3. Griping is an effective way to create awareness in order to accomplish an objective for the greater good

4. I challenge your presupposition that griping to bring awareness to a problem so that someone skilled can fix it is unproductive. It's called QA and it makes billions of productive dollars for business people that I mentioned in my second post above. It also makes open source software better. You're welcome.

Any other questions?

**Comment by DavidPesta** [comment]
@TPS - I have a question of my own for you. I haven't changed my purpose in this thread. My tone has also been more or less consistent from the beginning. After my first post you responded to me in complementary fashion with this:

"If you can believe it, this has already been filed many years ago...I really appreciate the re-filed issue, though, as this is 1 of my pet peeves, also!"

But then I posted a few more times with the same purpose in a similar tone to offer a reasonable defense (that others like Cilencia ended up agreeing with) and you send me this:

"why don't you fork & fix it yourself, rather than unproductively griping?"

What caused you to change your mind so dramatically about my purpose with this thread? Why was simply "filing the issue" commendable to you several months ago, and no longer acceptable?

Anyway, you did join my gripe by saying, "this is 1 of my pet peeves, also!" Therefore I'll be cooperative in my answer to your challenge: Why don't we both work together to figure out a way to get this fixed? Thank you for recruiting yourself to help me.

I'll start slowly looking at the source code and map out some ideas about where and what might be easily done to hack our keyboard to do something that helps. Pun intended I guess. I have no idea how long this will take us, but here it goes.

**Comment by meMongo** [comment]
I just skimmed over the comments so for give me if I'm repeating what someone else said.  This keyboard has the CONTROL KEY.  Use CNTL Z (duh) UNDO last.

**Comment by DavidPesta** [comment]
@meMongo - Ctrl-z doesn't work. Don't just skim the comments. Read them.

**Comment by Cilencia** [comment]
Yes, that's been mentioned a few times; it doesn't work in some apps because it is an app function, not a system function, and not universally supported.  Please feel free to read over the whole thread for further context.

**Comment by ThumpieBunnyEve** [comment]
DavidPesta, any progress? Where were you stumped during your source editing? Did you start a new thread for your custom apk file? Are you emailing them to requesties?  I've made changes to my "smart keyboard pro", having skinned it and tinkered with button alignment before, but I'm not an android programer yet. Basicly a script kiddy in the android scene. But I understand a little. Please publish your resarch progress here or in a new thread so we all can work together to forward this feature-fix into a dl'able state.

**Comment by DavidPesta** [comment]
@ThumpieBunnyEve - It's only been several days. I haven't had any time to put into it yet. I promised to start slowly looking into it, not quickly. :)

How about you and @TPS ? Anything yet?

I'll start posting in this thread as I get moments to look at it.

**Comment by DavidPesta** [comment]
I had another document blow up on my recently. This serves as a good reminder to keep coming back here and fiddle with it some more. Someday we'll get this fixed.

**Comment by ThumpieBunnyEve** [comment]
I cant recall if it was here or in another thread, but i read something about some experimental option some where you can tic- mark that will allow the control key to be in the bottom corner?  it might have been early in this thread... not sure.  but are you or anyone else here familiar with an option like that?

**Comment by TPS** [comment]
@ThumpieBunnyEve I put [detailed instructions earlier](https://github.com/klausw/hackerskeyboard/issues/487#issuecomment-201294229) w/ an [example screenshot](https://github.com/klausw/hackerskeyboard/issues/487#issuecomment-177280639).

**Comment by klausw** [comment]
Proposed fix in v1.39.0, I've just uploaded that to the alpha channel.

    Add Ctrl-A options, require Ctrl-Alt-A by default
    
    Add an option to configure "Ctrl-A" behavior to work around
    accidental text deletion when doing an unintended "select all"
    while typing.
    
    New default is to ignore Ctrl-A, and require use of Ctrl-Alt-A
    instead to do a "select all". Pressing Ctrl-A shows a toast message
    indicating this.
    
    The other available settings are to ignore Ctrl-A completely
    (with no toast message), or to enable Ctrl-A normally with no
    behavior override.
    


Use this link to opt-in to the alpha version: https://play.google.com/apps/testing/org.pocketworkstation.pckeyboard

It may take a few hours for the new version to propagate. Check your version at the bottom of the keyboard's settings menu.

**Comment by klausw** [comment]
@ThumpieBunnyEve @TPS FYI, as of alpha 1.39.1, the compact 5-row layout is now always offered as an option in the "Keyboard mode, portrait/landscape" settings, it is no longer necessary to unhide it separately.

**Comment by TPS** [comment]
@KlausW Thanks hugely!  This solves the problem brilliantly! :bow:

**Comment by DavidPesta** [comment]
@klausw - I want to thank you for your response and brilliant solution to this. Looking at your update, it would have taken me a **_very long time_** to figure that out. You are totally awesome for your work developing this keyboard! Even with my pain that I expressed in this discussion, the keyboard overall has been a great improvement to my workflow. Hero of heroes!

I also want to thank the community for putting up with my noise to bring attention to this. I know my rhetoric was uncomfortable and you gave me some flak for it, but I felt like it was necessary. How many years was this a problem? How many documents have you lost?

Again, what a wonderful solution! I am very thankful that this is now solved.

**Comment by klausw** [comment]
@DavidPesta FYI, I was seriously considering just shutting down the project during the course of this discussion[1]. Insulting volunteer developers is a risky strategy.

I do see your point that it's a serious bug and needed to be fixed, but I think comments such as "this is why business people rule the world and not developers" were not helpful. This is a free open source application, and the concept of a business strategy seems dubious in this context since there's no revenue.

For context, I wrote the keyboard back in 2011 when I got a brand-new Motorola Xoom and wanted a better input option than its default keyboard. I shared it since I thought others would find it useful also, and community contributions such as the many language-specific layouts have made it much better than I could have done on my own. (Though adapting layouts had been a lot of work in some cases, such as the French AZERTY keyboard where shift != caps, Turkish Iİıi, dead accents, RTL language support and more.) In exchange for putting in this work I get called an "illiterate retard" for misspelling "Portuguese" in release notes, and those are the moments when I wonder why I put up with the aggravation. It's been at the "works for me" stage for many years now.

[1] To be clear, shutting down the project would have been intended to make it clear to users that it's unsupported, and would have left open the option for others to build and publish their own versions based on the existing open source code which would still be available.

**Comment by DavidPesta** [comment]
@klausw - Point well taken. I was definitely more rough than I should have been with that comment and I apologize. I may have been cranky after losing an important document. That other comment you mentioned, made against you for simply misspelling "Portuguese" was plainly egregious. Totally uncalled for. I'm sad to see you put up with this stuff. Thank you for opening up like that.

Thank you also for your willingness to put in a little more support for a project that you no longer support. That you would do so for someone who insulted you is big of you. (Granted you saw the need for everyone, not just me.) Thank you again on behalf of all of us.

**Comment by ThumpieBunnyEve** [comment]
I feel a little bit noobish here, but i don't see an .apk file anywhere to download. i even opted in for the tester-version update, and then went to the googleplay page.   But it only updated to 1.38.2.1434  from 2015-04-21   the git-hub page doesn't have an apk file either. and i don't have an android programmer license, so i cant compile my own apk files.  (just tweek other peoples xml's in the directories they install to)  Just how do i acquire or update to version:  1.39.1  ??   my android is Marshmallow, and I'm on a Nexus 7, 2013 LTE model.  Is googlePlay just giving me the les updated testing version because of my model? Or am i stumbling over myself some other way?  Its not making itself obvious in spite of my best efforts.

**Comment by klausw** [comment]
@ThumpieBunnyEve I've uploaded the APK file here as a release: https://github.com/klausw/hackerskeyboard/releases/ 

I'll retire the Drive folder that's linked from [wiki/TestReleaseOptIn](https://github.com/klausw/hackerskeyboard/wiki/TestReleaseOptIn).

Has the tester opt-in worked for anyone? It's an automated process that I can't control directly, but the update should appear for you in the Play Store application some time after opting in, same as regular application updates.

FYI. you don't need an "Android programmer license" to build and install your own code, just the Android SDK. You only need to set up a developer (publisher) account if you want to offer an application on the Play Store, but that's a separate step. Compiling is a bit tricky for this application since it also needs the NDK, and since it predates the current Android Studio setup the instructions are out of date. Anyway, you shouldn't need to do this if you just want an unmodified release.

**Comment by ThumpieBunnyEve** [comment]
Well it did work for me to get the 1.38.1 when i opted in to the test version. So yes it worked. though i suspect persons using this keyboard are mostly using the stable version listed due to google plays advertisement of that version, and because other then the control-a issue, this is a pretty awesome human interface, many many features above and/or including, what others provide.  

I can still use voice with it, i can CtrlC  CtrlV  so long as the app I'm in permits such inputs. the stable version even with ctrl-a is quite spectacular.  harder for smaller phones and fatter fingers i suppose, but I'd rather the features be accessible on the first click, then having to tap a bunch of submenu keys to show pages of layouts. 

I recommend this keyboard to my friends.  Whom are a little leery only because of the name "Hackers Keyboard"  but once they start using it, they prefer it over their others.  I've tested a number of different ones, and keep coming back to this.  Some of the best ones, shove ad's in your face or pepper you with automatic notifications.  

Hackers keyboard may eventually outpace all the other types if it remains ad free, functional and feature packed as it is.  Even the stable version is advanced and useful enough that else on the market has to struggle with gimmicky skins in order to get installed. The tester-opt-in is just an afterthought for experienced users i think. 

Strong with the HID this one is, Yehhhs. ~says Yoda.

Future suggestions: 
~I've not checked to see if swipe-keys works, its not really my style of typing, but if not included, that is a popular textual input type people are using in abundance.
~In the settings menu it'd be peachy if it stated the version number presently installed.
~A skin-file selector/browser section might be fun for the keyboard artists out there. 
~The Shift-State key ideograms could be color-changeable (if they are Textually rendered rather then part of the png's of each button)  As black secondary's next to White primaries, on a #646464 background are a bit hard to see.

But the above, could simply be left open to skin-designers.

I'm using mostly the default look for all this so I've not dug too deeply.  Because, it's Good.  it's really useful and great as it is!  So i don't know if theres already a skins standardization framework or not. I've not diddled with any png's or xml's in this thing yet.  I'm pleased with the color scheme provided, and what I've selected from the settings menu from the prior installed versions that carried over.

Simply put, this is awesome and you've outdone hoards of other keyboard interfaces that I've tested out and tossed in the recycle-bin.  This is an input method for SuperUsers and it's hard to even stand anything else with how wonderfully spoiled i feel using the features offered in this app.  

To drive the point home, I have 2 bluetooth hardware keyboards and i still prefer HackersKeyboard app over both of them.  They just don't maintain the same features as your software!  So that's saying something!  ♥❤

**Comment by TPS** [comment]
🤷 I was so anxious to get updated, I just installed straight from the GDrive folder, so I've no feedback on whether Play has it.

@KlausW, that GDrive folder is often more accessible under Android than even the GitHub /releases, as I have it shared to *my* GDrive app.  If you don't mind, could you keep that folder share live?

**Comment by klausw** [comment]
@ThumpieBunnyEve I'm planning to distribute the new build as the new production version on the Play Store soon, but would like to collect some feedback first to make sure it doesn't cause regressions. The downside of having so many configurable options is that it's pretty much impossible to test all possible combinations, and I want to be careful not to break things for existing users.

Swiping key input is not supported and is not going to be - the underlying Gingerbread AOSP keyboard that the code is based on doesn't support it, and it's not really feasible to backport this. I think it would be easier to start fresh with a more current upstream AOSP keyboard and port the layouts and other features over, but I don't have any plans to do so. 

The Hacker's Keyboard settings menu does already show the installed version number all the way at the bottom, in the "Debugging" section.

Re theming - I'm not currently planning to work on this, but would be willing to merge contributions as long as they don't impact stability or greatly increase the package size. At this point there's no support for external theme packages, they need to be baked into the binary.

@TPS ok, I'll keep the Drive folder around also.

**Comment by TPS** [comment]
@KlausW @DavidPesta Any reason that this issue can't be <kbd>Closed</kbd>, as well?  Or are you waiting for the official release?

**Comment by klausw** [comment]
@TPS I prefer to keep it open and marked as "fixed in test" until the prod release. Main reason is that the default issue search skips closed issues, and that can be confusing for users of the prod version. This way the issue is still visible, and they can upgrade to the test version if appropriate.

The other bugs are showing as closed since I marked them as duplicates of this one.

**Comment by DavidPesta** [comment]
Every time I see "Ctrl-A is disabled. Use Ctrl-Alt-A instead, or enable it in settings." I will smile at another document that's been saved. But I will always remember that developer out there who was willing to go the extra mile and take the time to fix an important bug for the community long after that developer was done with the project. Thanks again @klausw.

**Comment by ThumpieBunnyEve** [comment]
Still using it, still loving it.  actually set a ctrl-Alt-A  function on my PC to match yours,  Not everything in windows (games mostly) knows how to select all, so i put a macro on my keyboard to that setup to mimic yours.   
::::::::start off topic::::::::::::::
my pc version is written in autohotkey, and it presses the following sequence, and has worked in every input so far except dos-box games @p@  it macros as follows,  (press ctrl-Home, press page-up(x2), hold Right'shift, press ctrl-End press page-down(x2), release Left'shift).   
::::::::end off topic:::::::::::::::

Also super glad the warning for ctrl A does not pop up in the way awaiting an "ok" dialog. Wee!

**Comment by 5uperstring** [comment]
I just installed your keyboard yesterday and just killed my second document 10 minutes ago. That is why I am here. The ctrl-A kill-all is an absolute disaster and I was astonished to find this long running thread about it with no fix!? Wtf is wrong with you? This is not a difficult fix. The only possible reason you haven't fixed it is because you always personally use a landscape tablet and so haven't nuked entire documents like everybody else who uses this app. When I get some free time I will fix my copy but since you clearly have no interest if fixing this you should at least put a big warning on the play store or change the name to "Doc destroying time bomb keyboard"

**Comment by DavidPesta** [comment]
@5uperstring - Please read this entire discussion before jumping to conclusions about the developer. It has been fixed and it is pending a push to full release, but you can enjoy the fix today by subscribing to the test version. I'd give you the link, but I want you to read the discussion.

We need to give the developer a lot of credit for sacrificing his time to provide the community with excellent software that he still chose to support for our benefit even after he has moved on. He did not have to fix this for us; we could have taken on the pain of figuring it out ourselves. So, while I can personally relate to the emotional pain of your loss (I expressed my own crankiness earlier), the anger here is very much misplaced. We need to be giving the developer our gratitude for bringing us this excellent keyboard, and even more gratitude for fixing it for us.

Again, please read the whole discussion and then subscribe to the fix.

**Comment by 5uperstring** [comment]
Why would I give the developer credit! Huh?! This thread is almost a year old and the release I installed yesterday killed two documents. Sheesh. The developer is getting no smiles and back pats from me. If I can fix his bug in an afternoon I don't see why it has been a year and he has yet to fix it. Sorry but if developing is only a part time weekend hobby then he should have put a warning on his play page last February to prevent countless documents from being nuked since then.

**Comment by TPS** [comment]
@5uperstring So, you came here POed & let off steam by having a infantile tantrum which abuses @klausw.  Congrats!  You earned my personal ban-hammer & probably [GitHub's](https://help.github.com/articles/github-community-guidelines/).  You certainly must be a treat IRL.  Since you obviously will ___NOT___ read the previous details in "[t]his thread [that']s almost a year old" before firing off ___2___ idiotic posts, I shan't bother honoring _you_ with a solution.  Sayanara & good luck with your future GitHub adventures after I report _you_.

@klausw Notwithstanding this ___moronic asshat troll___, for those of us silent faithful who have kept using your last release over the years, could the latest ß be promoted to release?  TIA! :bow:

**Comment by 5uperstring** [comment]
Give me a bread TPS. Glad you are having a group hug over here while his app is harming users. You don't get banned for pointing out that somebody is distributing malware. Sorry. However, you do get banned from the play store for distributing harmful software. I will fix this app by tomorrow and post it on the play store myself if klaus will allow it.

**Comment by DavidPesta** [comment]
@5uperstring - As happy as I am with the developer right now, I can sympathize with your pain and I understand your point. On the one hand, this is a volunteer project and the developer does not owe anybody anything. On the other hand, until the fix is pushed to production, the bug remains something that can hurt new people who do not expect this to happen. But putting everything into perspective, the developer did recognize this problem and did/is fixing it, so there isn't anything productive to complain about here anymore.

@TPS - Relax. This bug may have cost this fellow more than you think. It takes time to emotionally recover from the shock of this. It did for me, but I came around.

**Comment by 5uperstring** [comment]
Thank you David. The app did cost me a lot of time. It is no doubt that I am not alone. Everyone wants a full keyboard (especially with old Sparc Station Ctrl key placement! don't move that!) however, it should not be allowed to continue to harm users until the fixed version is ready to distribute. If the developer doesn't feel like fixing it then it would be taken down.

**Comment by TPS** [comment]
@all: @5uperstring has been reported for violating the above-linked abuse policy.  Keep it & civil & humane!

Also, it occurs to me the _original_ posts in this issue seem to set the same kind of tone as what we all just experienced (&, indeed, come mighty close to violating that abuse policy, _had it existed then_).  I'm going to go back & tone down my rhetoric whenever it's over-the-top.  I invite/welcome my fellow co-participants to do the same.

**Comment by 5uperstring** [comment]
TPS is the only one calling people names here. I did not call anyone any names. I pointed out facts and the potential consequences of his harmful software. I have developed numerous apps, distributed to people for free, with no ads and no profit for me. However, if I ever found out that one of my apps was doing this exact same thing to users I would immediately take it down until I could upload a non-harmful version.

**Comment by DavidPesta** [comment]
@TPS - Given the severity of the bug and the amount of damage that it has caused people, I do not think your reaction to people's emotions is rational. 5uperstring's words may have been abusive, but I understand his pain. We need to give him a break, give him a chance to cool down and apologize, as I have done. I am not going to justify his words, but I do not believe you are being fair to him by not recognizing his side of the issue. 5uperstring also reserves the right to report the app to the Google Play Store as malware, which would be equally irrational at this juncture. It is my hope that 5uperstring responds with more maturity by not doing so.

TPS, I think you are the one escalating the trouble here. Relax and give this fellow a chance to cool down.

**Comment by TPS** [comment]
Sorry, folks, I'm little sensitive about abuse of @klausw specifically because of [his detailed response re: his reaction _last_ time](https://github.com/klausw/hackerskeyboard/issues/487#issuecomment-255540845) which @5uperstring did, in fact, completely ignore.

However, @5uperstring, you're correct about the name-calling.  I'll use this amazing app to edit out.

**Comment by 5uperstring** [comment]
I won't be apologizing to the developer. As we sit here and discuss the issue documents are being destroyed. If you don't consider that malware then I am not sure how you are categorizing things. 
I understand that you guys feel very good to be helping to improve this nice keyboard and you have grown to like the guy who wrote it. But you also need to take a step back. If he is not very interested in fixing it then he should take it off google play and allow you guys to fix it yourselves and post the fixed version. You could even post TWO keyboard apps if you like: The first one identical to this one if you like! Just put a big warning on it that if you don't use ctrl-A frequently then you should instead install the other version. The second version simply has ctrl-A remapped to some harmless ctrl combination. 
Is there some reason this wasn't done? It keeps current happy people happy and stops future angry github posts from guys like me.

**Comment by TPS** [comment]
@5uperstring The fixed version has _already_ been posted to GPlay,  & this same issue tells you how to access it.

**Comment by DavidPesta** [comment]
@TPS - I love the developer of this app. But we cannot expect people to install the app and somehow know that they need to come to this discussion to apply the fixed test version to keep their documents from being destroyed. They're only going to know its a problem after its too late. We risk having more irate people coming here to demand answers. There's also no telling how many angry people there are who don't know about GitHub and never make it into this discussion. I want this to continue to be a successful app, not one that risks being reported as malware by someone like 5uperstring.

It would be super cool if this fix could be pushed over to the production "default installed" version ASAP so that we can put an end to this discussion. What's the holdup? Yes, I am a happy camper because it is fixed on my phone, but this really is an urgent hotfix for the community at large.

Again, I am a happy camper. But I would love this to finally make it into production.

**Comment by TPS** [comment]
@DavidPesta Where I'm irritated is that @5uperstring did indeed spend the time to find this discussion, ignored all of it (except maybe the title — maybe that could be edited to mention "fixed in ß" or some such), & did a drive-by [something] without getting the solution that's ___right here___!  Why bother?

**Comment by 5uperstring** [comment]
I didn't ignore the discussion. I know exactly what you guys are doing and have done. I don't like your particular solution but that is just personal preference. However, it is very easy to put two different apps on google play. One for emacs users and one for the rest of us. ;)

**Comment by DavidPesta** [comment]
@TPS - He was very very very very angry. Like I was. Look, he is better now. :)

@5uperstring - Any solution that no longer destroys my documents is gold to me. Everything else about this keyboard rocks!

@klausw - The world is ready for this update. Push push push! :D

**Comment by klausw** [comment]
Sheesh. Sorry @5uperstring and others who were affected by lost documents. I've now promoted 1.39.1 to production, Play Store should start pushing it out within the next day or so. I didn't get a chance earlier to sanity check it for crash reports or other potential issues, but so far I haven't seen problem reports for the alpha.

As a side note, if you're working on complex documents, I would strongly recommend using an editor that supports undo or preferably automatic snapshots to avoid lost work. There's more threats to your document than just malfunctioning keyboards by unhelpful developers, including application crashes, loss of network connectivity, and hardware failure. But that's a separate topic.

**Comment by klausw** [comment]
Closing, 1.39.1 is now set as live on Play Store. It may take a day or so until your device sees the update.

**Comment by 5uperstring** [comment]
Got the new version. Love it! You win. ... I apologize.

**Comment by cydhaselton** [comment]
Huge thanks for this update. Android tablet is my main dev system, so this will help tremendously.
